### PR TITLE
Feature/builder pickup to 1.11

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Analog to 1.10 pr

Review please

### CODE REVIEW

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 49 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 40 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 37 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L769](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L769): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L48](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L185](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L185): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L268](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L268): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Structure.java#L65](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L65): Reduce this class from 54 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Structure.java#L165](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L165): Remove this unused private "structureBlockEqualsWorldBlock" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/9107b39e9865bead0b26291e32e0c24ed637f808/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 1]
```diff
@@ -32,20 +35,60 @@
...
+     */
+    private static final float RANGE_HORIZONTAL_PICKUP = 45.0F;
+
+    /**
+     * Vertical range in which the lumberjack picks up items.
+     */
+    private static final float RANGE_VERTICAL_PICKUP   = 3.0F;
+
+    /**
+     * Number of ticks the lumberjack is standing still.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: Lumberjack ?

**Comment by xavierh** [comment]
Seems to be a lot of duplicated code between LumberJack and this (and maybe the Guard as well)
Could we not make this more generic and reusable ?

**Comment by Raycoms** [comment]
I actually copied all the lumberjack code in the AbstractEntityAIInteract
to reuse it.
But true, we should be even able to get some of the guard code there as
well.

2017-05-15 16:37 GMT-03:00 xavierh <notifications@github.com>:

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 2]
```diff
@@ -30,17 +29,15 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java]: MINOR Remove this unused import 'org.jetbrains.annotations.Nullable'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 50 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 41 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 37 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L769](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L769): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L48](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L185](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L185): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L268](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L268): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Structure.java#L65](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L65): Reduce this class from 54 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Structure.java#L165](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L165): Remove this unused private "structureBlockEqualsWorldBlock" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/f48276bb0f3fb72693f5ff08d128dda0fc73534d/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 3]
```diff
@@ -32,20 +35,60 @@
...
+     */
+    private static final int   STUCK_WAIT_TICKS        = 20;
+
+    /**
+     * The amount of time to wait while walking to items.
+     */
+    private static final int   WAIT_WHILE_WALKING      = 5;
+
+    /**
+     * Horizontal range in which the lumberjack picks up items.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: worker

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 4]
```diff
@@ -32,20 +35,60 @@
...
+     */
+    private static final float RANGE_HORIZONTAL_PICKUP = 45.0F;
+
+    /**
+     * Vertical range in which the lumberjack picks up items.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: worker

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 5]
```diff
@@ -32,20 +35,60 @@
...
+    private static final float RANGE_VERTICAL_PICKUP   = 3.0F;
+
+    /**
+     * Number of ticks the lumberjack is standing still.
+     */
+    private              int   stillTicks              = 0;
+
+    /**
+     * Used to store the path index
+     * to check if the lumberjack is still walking.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: worker

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 6]
```diff
@@ -32,20 +35,60 @@
...
+     */
+    private              int   previousIndex           = 0;
+
+    /**
+     * Positions of all items that have to be collected.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: If this is the items to be gathered, both the description and the name should mention it. Unless it was an AI solely to to the gathering

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 7]
```diff
@@ -225,4 +268,103 @@ private int getBlockMiningDelay(@NotNull final Block block, @NotNull final Block
                                       .getStrVsBlock(worker.getHeldItemMainhand(),
                                         block.getDefaultState())));
     }
+
+    public void fillItemsList()
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: same here
javadoc too

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 8]
```diff
@@ -225,4 +268,103 @@ private int getBlockMiningDelay(@NotNull final Block block, @NotNull final Block
...
+    {
+        items = searchForItems();
+    }
+
+    /**
+     * Search for all items around the Lumberjack.
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: worker

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 49 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 40 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 37 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L769](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L769): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L48](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L185](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L185): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L268](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L268): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Structure.java#L65](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L65): Reduce this class from 54 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Structure.java#L165](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L165): Remove this unused private "structureBlockEqualsWorldBlock" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/2fed832f6ecc333a201fe7ea3913858106379f8d/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 9]
```diff
@@ -225,4 +268,103 @@ private int getBlockMiningDelay(@NotNull final Block block, @NotNull final Block
...
+    {
+        items = null;
+    }
+
+    /**
+     * Get the items to gather list.
+     * @return a copy of it.
+     */
+    @Nullable
+    public List<BlockPos> getItems()
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: I find the name too generic

**Review by xavierh — CHANGES_REQUESTED** [review]
Some java doc need to be changed

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 49 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 40 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 37 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L769](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L769): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L48](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L185](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L185): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L268](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L268): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Structure.java#L65](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L65): Reduce this class from 54 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Structure.java#L165](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L165): Remove this unused private "structureBlockEqualsWorldBlock" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/d753096fdb78d077003d8b8e518f4afe3ab17829/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by xavierh — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 10]
```diff
@@ -147,20 +153,62 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
+        final BlockPos centerPos = currentStructure.getCenter();
+        if(centerPos.getY() == 0)
+        {
+            return Collections.emptyList();
+        }
+
+        AxisAlignedBB boundingBox
+                = new AxisAlignedBB(centerPos).expand(currentStructure.getLength() / 2.0, currentStructure.getHeight(), currentStructure.getWidth() / 2.0);
+
+        return world.getEntitiesWithinAABB(EntityItem.class, boundingBox)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: why duplicate this? Can we abstract that?

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
we could reduce this a bit :stuck_out_tongue_winking_eye:

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 49 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 40 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L58](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L58): class "AbstractEntityAIStructure" has 37 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L764](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L764): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L48](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L185](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L185): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L268](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L268): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Structure.java#L65](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L65): Reduce this class from 54 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Structure.java#L165](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L165): Remove this unused private "structureBlockEqualsWorldBlock" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/772960e6b397ba8f9cdd1dc43441348f3dadfbbc/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 10 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: done

**Review by marvin-bitterlich — APPROVED** [review]

**Pull request merged**